### PR TITLE
Move indices to InodeStore

### PR DESCRIPTION
**Pull request by ggezer**
https://github.com/Alluxio/alluxio/issues/8991

### CODE REVIEW

**Comment by ggezer** [comment]
@gpang PTAL. Thanks!

File: core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java [hunk 1]
```diff
@@ -233,6 +240,69 @@ public void close() {
     }
   }
 
+  @Override
+  public InodeIndice getIndice(InodeIndiceType indiceType) {
+    return new CachingStoreInodeIndice(indiceType);
+  }
+
+  /**
+   * Implementation of {@link alluxio.master.metastore.InodeStore.InodeIndice} for
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java]: NIT: this doesnt need to be fully qualified with the package name.

File: core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java [hunk 2]
```diff
@@ -233,6 +240,69 @@ public void close() {
...
+   * CachingInodeStore.
+   */
+  private class CachingStoreInodeIndice implements InodeIndice {
+    private InodeIndiceType mIndiceType;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java]: NIT: `final`

File: core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java [hunk 3]
```diff
@@ -233,6 +240,69 @@ public void close() {
...
+    }
+
+    @Override
+    public void clear() {
+      try {
+        // Flush indice cache.
+        mIndiceCache.flush();
+      } catch (InterruptedException ie) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException("Interrupting while flushing dirty indices to disk");
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java]: please include the index type in the message

File: core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java [hunk 4]
```diff
@@ -233,6 +240,69 @@ public void close() {
...
+    }
+
+    @Override
+    public int size() {
+      try {
+        // Flush indice cache.
+        mIndiceCache.flush();
+      } catch (InterruptedException ie) {
+        Thread.currentThread().interrupt();
+        throw new RuntimeException("Interrupting while flushing dirty indices to disk");
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java]: please include the index type in the message

File: core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java [hunk 5]
```diff
@@ -782,4 +852,53 @@ public void removeChild(String name) {
...
   }
+
+  /**
+   * Cache for inode indices.
+   *
+   * {@link Cache} is used for caching writes to the backing store.
+   * Indices are defined by a composite key of <InodeIndice, InodeId>. Value is
+   * the InodeId itself.
+   *
+   * This cache is not used for reads because:
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/caching/CachingInodeStore.java]: Could this be an issue? Each worker queries for the pinlist every second. That would mean every worker will query for the list every second?

File: core/server/master/src/main/java/alluxio/master/metastore/heap/HeapInodeStore.java [hunk 6]
```diff
@@ -48,6 +54,18 @@
   // Map from inode id to ids of children of that inode. The inner maps are ordered by child name.
   private final TwoKeyConcurrentMap<Long, String, Long, Map<String, Long>> mEdges =
       new TwoKeyConcurrentMap<>(() -> new ConcurrentHashMap<>(4));
+  /** Map for storing indices. */
+  private HashMap<InodeIndiceType, Set<Long>> mIndices;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/heap/HeapInodeStore.java]: NIT: `final` ?

File: core/server/master/src/main/java/alluxio/master/metastore/heap/HeapInodeStore.java [hunk 7]
```diff
@@ -69,6 +87,50 @@ public void removeChild(long parentId, String name) {
...
+  @Override
+  public InodeIndice getIndice(InodeIndiceType indiceType) {
+    return new HeapStoreInodeIndice(mIndices.get(indiceType));
+  }
+
+  /**
+   * Implementation of {@link alluxio.master.metastore.InodeStore.InodeIndice} for HeapInodeStore.
+   */
+  private class HeapStoreInodeIndice implements InodeIndice {
+    private Set<Long> mIndicedIds;
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/heap/HeapInodeStore.java]: NIT: `final` ?

File: core/server/master/src/main/java/alluxio/master/metastore/heap/HeapInodeStore.java [hunk 8]
```diff
@@ -128,24 +193,68 @@ public void clear() {
...
       if (Thread.interrupted()) {
         throw new InterruptedException();
       }
-      inode.toProto().writeDelimitedTo(output);
+      inode.toProto().writeDelimitedTo(inodesOutput);
     }
+    chunked.endChunks();
+    // Write indices
+    for (InodeIndiceType indice : InodeIndiceType.values()) {
+      String indiceCheckpointName = "HEAP_INODE_STORE_INDICES_" + indice.name();
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/metastore/heap/HeapInodeStore.java]: "HEAP_INODE_STORE_INDICES_" should be a constant

**Pull request closed without merging**